### PR TITLE
[Perf] fuse all_to_all_4D

### PR DESCRIPTION
**Pull request by ZJY0516**
## Purpose
Ulysses SP swaps a tensor from seq-sharded to head-sharded (and back) around attention via an all-to-all. The usual path (`all_to_all_4D`) is three steps: `permute` (copy) → NCCL `all_to_all_single` → `permute` (copy).

`all_to_all_permute` (ported from [pytorch#178230](https://github.com/pytorch/pytorch/pull/178230)) fuses all three into one kernel: the input lives in a **NCCL symmetric-memory window**, and each rank reads its slice **directly from every peer's memory at the right strided offset**, writing the target layout as it goes (`ncclGetLsaPointer` + LSA acquire/release barriers). The strided remote read *is* the permute.

**Why faster:** no permute copies, one kernel launch over all peers, one-sided NVLink reads instead of NCCL's general all-to-all — **1.5–4× faster** than `all_to_all_4D` on GB200 on the exchange itself.

Directions: `(scatter=1, gather=0)` forward (scatter heads, gather seq), `(scatter=0, gather=1)` reverse. Requires the input copied into the symm-mem window and NCCL ≥ 2.28.

## Test Plan

<details>
<summary> test code </summary>

```
"""E2E serving A/B for the fused Ulysses all-to-all (VLLM_OMNI_ULYSSES_A2A_PERMUTE).

For each mode (baseline vs fused) it launches `vllm serve` with the flag set,
sends a few Qwen-Image requests to /v1/images/generations, and reports the median
end-to-end request latency. Generated images are saved so you can diff them.

    python test_a2a_permute_qwen.py

"""

import base64
import os
import signal
import statistics
import subprocess
import time

import requests

MODEL = "Qwen/Qwen-Image"
PORT = 8091
USP = 4          # ulysses degree (GPUs)
H = W = 1536
STEPS = 10
SEED = 42
WARMUP, MEASURE = 1, 3
PROMPT = "a cup of coffee on a wooden table, morning light"
URL = f"http://localhost:{PORT}"


def wait_until_ready(proc: subprocess.Popen, timeout: float = 900) -> None:
    deadline = time.time() + timeout
    while time.time() < deadline:
        if proc.poll() is not None:
            raise SystemExit("server exited before becoming ready")
        try:
            if requests.get(f"{URL}/v1/models", timeout=2).status_code == 200:
                return
        except requests.RequestException:
            pass
        time.sleep(3)
    raise SystemExit("server did not become ready in time")


def generate() -> tuple[float, bytes]:
    payload = {
        "prompt": PROMPT,
        "response_format": "b64_json",
        "n": 1,
        "size": f"{W}x{H}",
        "num_inference_steps": STEPS,
        "true_cfg_scale": 4.0,
        "seed": SEED,
    }
    t0 = time.perf_counter()
    r = requests.post(f"{URL}/v1/images/generations", json=payload, timeout=600)
    r.raise_for_status()
    latency_ms = (time.perf_counter() - t0) * 1000
    img = base64.b64decode(r.json()["data"][0]["b64_json"])
    return latency_ms, img


def run_mode(mode: str) -> float:
    """Launch the server with the flag set, time MEASURE requests, save the image."""
    env = {**os.environ, "VLLM_OMNI_ULYSSES_A2A_PERMUTE": mode, "VLLM_WORKER_MULTIPROC_METHOD": "spawn"}
    proc = subprocess.Popen(
        ["vllm", "serve", MODEL, "--omni", "--port", str(PORT), "--usp", str(USP)],
        env=env,
    )
    try:
        wait_until_ready(proc)
        for _ in range(WARMUP):
            generate()
        latencies, img = [], None
        for _ in range(MEASURE):
            ms, img = generate()
            latencies.append(ms)
        with open(f"/tmp/qwen_serve_{mode}.png", "wb") as f:
            f.write(img)
        return statistics.median(latencies)
    finally:
        proc.send_signal(signal.SIGINT)
        try:
            proc.wait(timeout=90)
        except subprocess.TimeoutExpired:
            proc.kill()
        time.sleep(3)  # let the port free up before the next mode


def main() -> None:
    latency = {}
    for mode, name in (("0", "baseline"), ("1", "fused")):
        print(f"\n=== {name} (VLLM_OMNI_ULYSSES_A2A_PERMUTE={mode}) ===", flush=True)
        latency[mode] = run_mode(mode)
        print(f"  {name}: median e2e latency = {latency[mode]:.1f} ms", flush=True)

    import numpy as np
    from PIL import Image

    a = np.asarray(Image.open("/tmp/qwen_serve_0.png")).astype(int)
    b = np.asarray(Image.open("/tmp/qwen_serve_1.png")).astype(int)
    max_diff = int(np.abs(a - b).max())
    speedup = latency["0"] / latency["1"]

    print("\n================ RESULT ================")
    print(f"baseline e2e latency : {latency['0']:.1f} ms")
    print(f"fused    e2e latency : {latency['1']:.1f} ms")
    print(f"speedup              : {speedup:.2f}x ({(1 - 1 / speedup) * 100:+.1f}%)")
    print(f"output identical     : {max_diff == 0} (max abs pixel diff = {max_diff})")
    print("========================================")


if __name__ == "__main__":
    main()

```
</details>

## Test Result

step 50 B300

```
================ RESULT ================
baseline e2e latency : 7452.0 ms
fused    e2e latency : 9452.9 ms
speedup              : 0.79x (-26.9%)
output identical     : True (max abs pixel diff = 0)
========================================
```

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
**BLOCKING:**

- **Test Coverage** — This PR exceeds 1000 LOC. Please run the [L3 tests](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/test_guide/#l3-level--l4-level) locally and paste the results in the PR description.

Performance optimization looks solid, but we need L3 test evidence for changes of this size.

File: vllm_omni/diffusion/distributed/a2a_permute.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    symm_shape: tuple[int, ...], dtype: torch.dtype, device: torch.device, group_name: str
+) -> torch.Tensor:
+    """Get (or lazily allocate + rendezvous) a symmetric-memory buffer.
+
+    Rendezvous is a collective; on a cache miss every rank hits the same shape in
+    lockstep (SP is symmetric), so they rendezvous together. A one-element
+    all_reduce first guarantees the group's NCCL communicator exists (required
+    before NCCL-symm rendezvous, else it segfaults).
+    """
+    key = (symm_shape, dtype, group_name)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/distributed/a2a_permute.py:82]: This cache returns a single symmetric-memory staging buffer for each `(shape, dtype, group_name)`, and later calls overwrite it with `symm_in.copy_()` before launching the exchange. That is only safe if forwards are strictly serialized on one stream. In serving, overlapping requests/replays or multiple CUDA streams with the same shape can race: one call can overwrite the buffer while an earlier `all_to_all_permute` kernel is still reading it, causing silent output corruption. Please either use per-stream/request scratch buffers, add stream-event lifetime management, or enforce/document serialized execution when this path is enabled.

File: vllm_omni/diffusion/distributed/csrc/a2a_permute.cu [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+    copy_bytes_vec16_aligned(
+        reinterpret_cast<const char*>(src_row),
+        reinterpret_cast<char*>(dst_row),
+        copy_row_bytes, coop.thread_rank(), coop.size());
+  }
+  bar.sync(coop, cuda::memory_order_release);
+}
+
+// ---- own devcomm cache (replaces PR's keyed manager API) --------------------
+ncclDevComm get_or_create_devcomm(ncclComm_t comm, const std::string& group_name) {
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/distributed/csrc/a2a_permute.cu:102]: This devcomm cache is keyed only by `group_name`, but the `ncclDevComm` is created from a specific `ncclComm_t`/device. If the same group name is reused after communicator recreation, or in a multi-device process, this can return a devcomm for the wrong communicator/device. Please include device and communicator identity in the cache key, or add a hard assertion that this process model cannot hit those cases.

File: vllm_omni/diffusion/attention/parallel/ulysses.py [hunk 3]
```diff
@@ -329,9 +335,21 @@ def pre_attention(
...
 
             # (bs, seq_len/P, head_cnt, head_size) -> (bs, seq_len, head_cnt/P, head_size)
-            query = SeqAllToAll4D.apply(self._ulysses_pg, query, self._scatter_idx, self._gather_idx, self._use_sync)
-            key = SeqAllToAll4D.apply(self._ulysses_pg, key, self._scatter_idx, self._gather_idx, self._use_sync)
-            value = SeqAllToAll4D.apply(self._ulysses_pg, value, self._scatter_idx, self._gather_idx, self._use_sync)
+            if _a2a_permute_enabled(self._scatter_idx, self._gather_idx, ulysses_world_size):
+                from vllm_omni.diffusion.distributed.a2a_permute import ulysses_qkv_fwd
+
+                gname = self._ulysses_pg.group_name
+                query = ulysses_qkv_fwd(query, gname, ulysses_world_size)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/attention/parallel/ulysses.py:342]: The fused path does not preserve the existing `use_sync` behavior. The baseline `SeqAllToAll4D.apply(..., self._use_sync)` synchronizes when requested, but `ulysses_qkv_fwd` has no equivalent parameter and the reverse fused calls have the same issue. Even if the default is usually false, enabling this env var should not silently change the synchronization contract.

File: vllm_omni/diffusion/distributed/a2a_permute.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    global _BUILT
+    if _BUILT:
+        return
+    with _BUILD_LOCK:
+        if _BUILT:
+            return
+        from torch.utils.cpp_extension import load
+
+        here = os.path.dirname(__file__)
+        src = os.path.join(here, "csrc", "a2a_permute.cu")
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/distributed/a2a_permute.py:50]: Please add an installed-wheel smoke test for this enabled path. The feature JIT-loads `csrc/a2a_permute.cu` from the package at runtime and depends on NCCL symmetric-memory headers/libs being discoverable in the installed environment, so a source checkout test alone can miss packaging/runtime failures.

**Review by hsliuustc0106 — COMMENTED** [review]
I left a few inline comments on the main correctness/performance risks I found. The optimization direction is promising, but I think these need to be resolved or explicitly constrained before enabling the fused path.

**Comment by baonudesifeizhai** [comment]
after update with upstream 

 
Metric | Baseline | Fused | Improvement
-- | -- | -- | --
Denoising throughput | 2.31 it/s | 4.39 it/s | 1.90×
Denoising latency (approx.) | 4.34 s | 2.28 s | 47.5% lower
Pipeline latency | 11.71 s | 8.33 s | 28.9% lower
End-to-end latency | 15.1922 s | 12.5812 s | 17.2% lower, 1.21× faster
Peak GPU memory | 24.06 GiB | 23.28 GiB | 796 MiB lower

 

Cosmos3-Edge 
4 card b300  
run_381_compile() {
  local fused="$1"
  local name="$2"

  VLLM_OMNI_ULYSSES_A2A_PERMUTE="$fused" \
  python examples/offline_inference/text_to_video/text_to_video.py \
    --model /mnt/minecraft/model/Cosmos3-Edge \
    --model-class-name Cosmos3OmniDiffusersPipeline \
    --prompt "A robot arm is cleaning a plate in the kitchen." \
    --height 480 \
    --width 832 \
    --num-frames 381 \
    --fps 24 \
    --num-inference-steps 10 \
    --guidance-scale 5.0 \
    --flow-shift 3.0 \
    --seed 42 \
    --ulysses-degree 4 \
    --ring-degree 1 \
    --extra-body '{"guardrails":false,"max_sequence_length":16384,"use_resolution_template":false,"use_duration_template":false}' \
    --output "/tmp/cosmos3_edge_381_compile_${name}.mp4" \
    2>&1 | tee "/tmp/cosmos3_edge_381_compile_${name}.log"
}
